man/cwtemplate.1
\"
\"  Copyright 2011-2014 James Hunt <james@jameshunt.us>
\"
\"  This file is part of Clockwork.
\"
\"  Clockwork is free software: you can redistribute it and/or modify
\"  it under the terms of the GNU General Public License as published by
\"  the Free Software Foundation, either version 3 of the License, or
\"  (at your option) any later version.
\"
\"  Clockwork is distributed in the hope that it will be useful,
\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
\"  GNU General Public License for more details.
\"
\"  You should have received a copy of the GNU General Public License
\"  along with Clockwork.  If not, see <http://www.gnu.org/licenses/>.
\"

.TH cwa "1" "September 2013" "Clockwork" "Clockwork Template Utility"
.SH NAME
.PP
cwtemplate \- Clockwork Template Utility

.SH SYNOPSIS
.PP
.B cwtemplate
template-file > out-file
.br
.B cwtemplate
-c /path/to/cwa.conf template-file

.SH DESCRIPTION
.PP
.B Clockwork
is a system configuration management system designed to securely enforce
policies on one or more host systems.  It can ensure that files have the
prescribed attributes (owner, group, permissions, etc.) and content.  It
can maintain a specific set of installed packages, and even make sure that
system user accounts exist.
.PP
.B cwtemplate
is a utility program to aide in the creation of fact-based file templates.
It renders a single template against various sets of facts, either gathered
from the running system, specified in a cache file, or set via command-line
arguments.

.SH OPTIONS
.PP
\fB\-V\fR, \fB\-\-version\fR
.RS 8
Print version and copyright information.
.RE

\fB\-c\fR, \fB\-\-config\fR \fI/path/to/cwa.conf\fR
.RS 8
Specify the path to an alternate configuration file.  If not given, the
default configuration file (/etc/clockwork/cwa.conf) is used instead.
.RE

\fB\-F\fR, \fB\-\-facts\fR \fI/path/to/cached.facts\fR
.RS 8
Tell \fBcwtemplate\fR to read facts from a file, where each line is a fact,
and takes the form of \fBkey=values\\n\fR.  If specified, gathered facts
are ignored.
.RE

\fB\-f\fR, \fB\-\-fact\fR \fIfact=value\fR
.RS 8
Override a single fact with a new value.  This can be used with either
gathered or cached facts; command-line facts always take precedence.
.RE

\fB\-C\fR, \fB\-\-check\fR
.RS 8
Don't render the template, just check that the template syntax is correct.
.RE

.SH EXIT STATUS
.PP
.B cwtemplate
exits 0 if it was able to read fact data and parse the template.

.SH FILES
.PP
.TP 8
.B /etc/clockwork/cwa.conf
Default configuration file that
.B cwa
reads on start up.  See \fBcwa.conf\fR(5) for details.

.SH SEE ALSO
.PP
\fBclockwork\fR(7), \fBcwa.conf\fR(5)

.SH AUTHOR
.PP
Clockwork was designed and written by James Hunt.


src/cwtemplate.c
/*
  Copyright 2011-2014 James Hunt <james@jameshunt.us>

  This file is part of Clockwork.

  Clockwork is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Clockwork is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Clockwork.  If not, see <http://www.gnu.org/licenses/>.
 */

#define CLIENT_SPACE

#include "clockwork.h"

#include <getopt.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/time.h>

#include "client.h"
#include "template.h"

static struct cwt_client* cwt_options(int argc, char **argv);
static void show_version(void);
static void show_help(void);

struct cwt_client {
	int check_only;   /* don't render, just verify */
	char *template;   /* file to render */
	char *facts_from; /* where to read facts; NULL = standard */
	struct hash *facts;

	int log_level;
	char *config_file;
};

/**************************************************************/

int main(int argc, char **argv)
{
	struct cwt_client *cwt;
	struct client *c;
	struct template *tpl;

	cwt = cwt_options(argc, argv);
	c = xmalloc(sizeof(struct client));
	c->log_level   = cwt->log_level;
	c->config_file = cwt->config_file;

	if (client_options(c) != 0) {
		fprintf(stderr, "Unable to process client options");
		exit(2);
	}
	c->log_level = log_set(c->log_level);
	INFO("Log level is %s (%u)", log_level_name(c->log_level), c->log_level);

	INFO("Gathering facts");
	c->facts = hash_new();
	if (cwt->facts_from) {
		INFO("Reading cached facts from %s", cwt->facts_from);
		if (fact_cat_read(cwt->facts_from, c->facts) != 0) {
			CRITICAL("Unable to read cached facts");
			exit(1);
		}
	} else if (fact_gather(c->gatherers, c->facts) != 0) {
		INFO("Gathering facts from %s", c->gatherers);
		CRITICAL("Unable to gather facts");
		exit(1);
	}
	if (cwt->facts) {
		INFO("Merging collected facts with command-line overrides");
		hash_merge(c->facts, cwt->facts);
	}

	tpl = template_create(cwt->template, c->facts);
	if (!tpl) {
		CRITICAL("Failed to parse template");
		exit(1);
	}
	if (cwt->check_only) {
		template_render(tpl);
		fprintf(stderr, "%s OK\n", cwt->template);
	} else {
		printf("%s\n", template_render(tpl));
	}

	return 0;
}

/**************************************************************/

static struct cwt_client* cwt_options(int argc, char **argv)
{
	struct cwt_client *cwt;

	const char *short_opts = "h?c:s:p:nvqF:f:DVC";
	struct option long_opts[] = {
		{ "help",     no_argument,       NULL, 'h' },
		{ "config",   required_argument, NULL, 'c' },
		{ "verbose",  no_argument,       NULL, 'v' },
		{ "quiet",    no_argument,       NULL, 'q' },
		{ "facts",    required_argument, NULL, 'F' },
		{ "fact",     required_argument, NULL, 'f' },
		{ "version",  no_argument,       NULL, 'V' },
		{ "check",    no_argument,       NULL, 'C' },
		{ 0, 0, 0, 0 },
	};

	int opt, idx = 0;

	cwt = xmalloc(sizeof(struct cwt_client));

	while ( (opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != -1) {
		switch(opt) {
		case 'h':
		case '?':
			show_help();
			exit(0);
		case 'V':
			show_version();
			exit(0);
		case 'c':
			free(cwt->config_file);
			cwt->config_file = strdup(optarg);
			break;
		case 'v':
			cwt->log_level++;
			break;
		case 'q':
			cwt->log_level--;
			break;
		case 'F':
			free(cwt->facts_from);
			cwt->facts_from = strdup(optarg);
			break;
		case 'f':
			if (!cwt->facts) {
				cwt->facts = hash_new();
			}
			if (fact_parse(optarg, cwt->facts) != 0) {
				perror("fact_parse");
				exit(1);
			}
			break;
		case 'C':
			cwt->check_only = 1;
			break;
		}
	}

	if (optind > 0 && optind < argc) {
		cwt->template = strdup(argv[optind]);
	}

	return cwt;
}

static void show_version(void)
{
	printf("cwtemplate (Clockwork) " VERSION "\n"
	       "Copyright 2011-2014 James Hunt\n");
}

static void show_help(void)
{
	printf("USAGE: cwtemplate [OPTIONS]\n"
	       "\n"
	       "  -h, --help            Show this helpful message.\n"
	       "                        (for more in-depth help, check the man pages.)\n"
	       "\n"
	       "  -V, --version         Print version and copyright information.\n"
	       "\n"
	       "  -c, --config          Specify the path to an alternate configuration file.\n"
	       "\n"
	       "  -C, --check           Don't render the template, just check the syntax.\n"
	       "\n"
	       "  -F, --facts           Specify the path to a cached facts file.\n"
	       "\n"
	       "  -f, --fact K=V        Set a fact.  Can be given more than once.\n"
	       "\n");
}
